src/Renci.SshNet/Security/Cryptography/Ciphers/AesCipher.CtrImpl.cs

```diff
@@ -0,0 +1,230 @@
+
+                _packedIV = GetPackedIV(iv);
+            }
+
+            public override byte[] Encrypt(byte[] input, int offset, int length)
+            {
+                return CTREncryptDecrypt(input, offset, length);
+            }
+
+            public override byte[] Decrypt(byte[] input, int offset, int length)
+            {
+                return CTREncryptDecrypt(input, offset, length);
+            }
+
+            public override int DecryptBlock(byte[] inputBuffer, int inputOffset, int inputCount, byte[] outputBuffer, int outputOffset)
+            {
+                throw new NotImplementedException($"Invalid usage of {nameof(DecryptBlock)}.");
+            }
+
+            public override int EncryptBlock(byte[] inputBuffer, int inputOffset, int inputCount, byte[] outputBuffer, int outputOffset)
+            {
+                throw new NotImplementedException($"Invalid usage of {nameof(EncryptBlock)}.");
+            }
+
+            private byte[] CTREncryptDecrypt(byte[] data, int offset, int length)
+            {
+                var count = length / BlockSize;
+                if (length % BlockSize != 0)
+                {
+                    count++;
+                }
+
+                var buffer = new byte[count * BlockSize];
+                CTRCreateCounterArray(buffer);
+                _ = _encryptor.TransformBlock(buffer, 0, buffer.Length, buffer, 0);
+                ArrayXOR(buffer, data, offset, length);
+
+                // adjust output for non-blocksized lengths
+                if (buffer.Length > length)
+                {
+                    Array.Resize(ref buffer, length);
+                }
+
+                return buffer;
+            }
+
+#if NETSTANDARD2_1_OR_GREATER || NET6_0_OR_GREATER
+
+            // creates the Counter array filled with incrementing copies of IV
+            private void CTRCreateCounterArray(byte[] buffer)
+            {
+                var counter = MemoryMarshal.Cast<byte, uint>(buffer.AsSpan());
+
+                // fill array with IV, increment by 1 for each copy
+                var len = counter.Length;
+                for (var i = 0; i < len; i += 4)
+                {
+                    counter[i] = _packedIV[0];
+                    counter[i + 1] = _packedIV[1];
+                    counter[i + 2] = _packedIV[2];
+                    counter[i + 3] = _packedIV[3];
+
+                    // increment IV (little endian)
+                    if (_packedIV[3] < 0xFF000000u)
+                    {
+                        _packedIV[3] += 0x01000000u;
+                    }
+                    else
+                    {
+                        uint j = 3;
+                        do
+                        {
+                            _packedIV[j] = SwapEndianness(SwapEndianness(_packedIV[j]) + 1);
+                        }
+                        while (_packedIV[j] == 0 && --j >= 0);
+                    }
+                }
+            }
```

Comment:
The `_packedIV` logic looks unnecessarily complicated. I would be surprised if it is any better than simply:

```suggestion
            private static byte[] CTRCreateCounterArray(byte[] iv, int blocks)
            {
                Debug.Assert(iv.Length == 16);
                
                var counter = new byte[blocks * 16];
                
                for (int i = 0; i < blocks; i++)
                {
                    Buffer.BlockCopy(iv, 0, counter, i * 16, 16);
            
                    var j = 16;
                    while (--j >= 0 && ++iv[j] == 0)
                    {
                        // Intentionally empty block
                    }
                }
                
                return counter;
            }
```

Which would work on all platforms. This is not blocking feedback.

Reply:
Current code:
| Method      | Mean     | Error    | StdDev   | Gen0   | Allocated |
|------------ |---------:|---------:|---------:|-------:|----------:|
| Encrypt_CTR | 12.39 us | 0.245 us | 0.319 us | 3.9978 |   32.8 KB |
| Decrypt_CTR | 12.52 us | 0.249 us | 0.417 us | 3.9978 |   32.8 KB |

Your proposed code:
| Method      | Mean     | Error    | StdDev   | Gen0   | Allocated |
|------------ |---------:|---------:|---------:|-------:|----------:|
| Encrypt_CTR | 22.93 us | 0.449 us | 0.480 us | 3.9978 |   32.8 KB |
| Decrypt_CTR | 22.32 us | 0.290 us | 0.257 us | 3.9978 |   32.8 KB |

I had already tried that, it's the naive implementation. The 4.6 fallback code is an optimization of that, and the Net7 code is twice as fast. Blockcopy is actually slow, working with int32 is much faster. I also tested int64, there's no significant difference.

Reply:
Fair enough

Reply:
1. It seems like the `else` block does not get hit in the AES tests. Could you please add one which does? e.g. using an IV of 0xFF..FF

2. The endianness logic needs to cater for big-endian systems, i.e. using [BitConverter.IsLittleEndian](https://learn.microsoft.com/en-us/dotnet/api/system.bitconverter.islittleendian). Something like

```c#
_packedIV[j] = BitConverter.IsLittleEndian
    ? SwapEndianness(SwapEndianness(_packedIV[j]) + 1)
    : _packedIV[j] + 1;
```
And the same for the `if` block above.

Alternatively... https://github.com/Rob-Hague/SSH.NET/commit/25172b241cf706c5f3190366731d977eda52003a

Reply:
Are you telling me that the rest of SSH.Net works on Big Endian ARM v6 or others? I find that unlikely, given lots of code that fills in packet structures by simply assigning an int32. There are no tests for that, except for a PackTest that tests conversion functions that are mostly only used in that test class. Nowadays even ARM is little endian by default.

Reply:
> Alternatively... [Rob-Hague@25172b2](https://github.com/Rob-Hague/SSH.NET/commit/25172b241cf706c5f3190366731d977eda52003a)

Please provide performance numbers for this change.

Reply:
> 1. It seems like the `else` block does not get hit in the AES tests. Could you please add one which does? e.g. using an IV of 0xFF..FF

Please provide a PR for this.

Reply:
> Are you telling me that the rest of SSH.Net works on Big Endian ARM v6 or others? I find that unlikely, given lots of code that fills in packet structures by simply assigning an int32.

Could you please provide an example? On a cursory glance I could not find any offenders.

I don't know if the library works on big endian architecture, but I know that this code doesn't, so why not get it right?

> Please provide performance numbers for this change.

```

BenchmarkDotNet v0.13.9+228a464e8be6c580ad9408e98f18813f6407fb5a, Windows 11 (10.0.22621.2715/22H2/2022Update/SunValley2)
AMD Ryzen 7 5700U with Radeon Graphics, 1 CPU, 16 logical and 8 physical cores
.NET SDK 8.0.100
  [Host]     : .NET 7.0.14 (7.0.1423.51910), X64 RyuJIT AVX2
  DefaultJob : .NET 7.0.14 (7.0.1423.51910), X64 RyuJIT AVX2


```
d34f521b4acfe6baa02e53a
| Method      | Mean     | Error    | StdDev   | Gen0    | Allocated |
|------------ |---------:|---------:|---------:|--------:|----------:|
| Encrypt_CTR | 11.66 μs | 0.125 μs | 0.111 μs | 15.8691 |   32.8 KB |
| Decrypt_CTR | 11.66 μs | 0.088 μs | 0.078 μs | 15.8691 |   32.8 KB |

25172b241cf706c5f3190366
| Method      | Mean     | Error     | StdDev    | Gen0    | Allocated |
|------------ |---------:|----------:|----------:|--------:|----------:|
| Encrypt_CTR | 9.179 μs | 0.1522 μs | 0.1424 μs | 15.8691 |  32.77 KB |
| Decrypt_CTR | 8.909 μs | 0.0592 μs | 0.0462 μs | 15.8691 |  32.77 KB |

Reply:
Nice performance, I'll merge the code.

> I don't know if the library works on big endian architecture, but I know that this code doesn't, so why not get it right?

Because the level of nitpicking on this PR is already absurd. You know you can add another PR once this one is merged right?


> Could you please provide an example? On a cursory glance I could not find any offenders.

This [here](https://github.com/sshnet/SSH.NET/blob/develop/src/Renci.SshNet/Messages/Message.cs#L107) calls the stream.Write method [here](https://github.com/sshnet/SSH.NET/blob/develop/src/Renci.SshNet/Common/SshDataStream.cs#L65) which [converts to BigEndian](https://github.com/sshnet/SSH.NET/blob/develop/src/Renci.SshNet/Common/Pack.cs#L174) without checking the current arch.

Reply:
> This [here](https://github.com/sshnet/SSH.NET/blob/develop/src/Renci.SshNet/Messages/Message.cs#L107) calls the stream.Write method [here](https://github.com/sshnet/SSH.NET/blob/develop/src/Renci.SshNet/Common/SshDataStream.cs#L65) which [converts to BigEndian](https://github.com/sshnet/SSH.NET/blob/develop/src/Renci.SshNet/Common/Pack.cs#L174) without checking the current arch.

AFAIU this code will work fine. Bitwise operations are endianness-agnostic.

I'll let @WojciechNagorski finish off this PR.

Reply:
> AFAIU this code will work fine. Bitwise operations are endianness-agnostic.

0x12345678 -> reverse = 0x78563412. When you then store it in memory, a little endian CPU will write 12 34 56 78 (as intended by SSH), but a big endian CPU will now write 78 56 34 12. To obtain the same result the BigEndian CPU should not reverse endianness.